File: 课程作业/C、C++/data_structure/test/dbc.cpp
#include <iostream>
#include <string>
#include <windows.h>
#include <sqltypes.h>
#include <sql.h>
#include <sqlext.h>

using namespace std;

// SQL Server连接信息
#define SQL_SERVER "your_server_name"
#define SQL_USER "your_username"
#define SQL_PASSWORD "your_password"
#define SQL_DB "UserDB"

// 函数声明
void registerUser(SQLHANDLE hStmt, const string& username, const string& password);
bool loginUser(SQLHANDLE hStmt, const string& username, const string& password);

int main() {
    // 连接数据库
    SQLHANDLE hEnv;
    SQLHANDLE hConn;
    SQLHANDLE hStmt;

    SQLAllocHandle(SQL_HANDLE_ENV, SQL_NULL_HANDLE, &hEnv);
    SQLSetEnvAttr(hEnv, SQL_ATTR_ODBC_VERSION, (SQLPOINTER)SQL_OV_ODBC3, 0);
    SQLAllocHandle(SQL_HANDLE_DBC, hEnv, &hConn);

    SQLCHAR* connectionString = (SQLCHAR*)"DRIVER={SQL Server};SERVER=" SQL_SERVER ";DATABASE=" SQL_DB ";UID=" SQL_USER ";PWD=" SQL_PASSWORD ";";
    SQLRETURN ret = SQLDriverConnect(hConn, NULL, connectionString, SQL_NTS, NULL, 0, NULL, SQL_DRIVER_NOPROMPT);

    // 检查连接是否成功
    if (!SQL_SUCCEEDED(ret)) {
        cout << "Failed to connect to SQL Server." << endl;
        return 1;
    }

    // 创建用户注册/登录界面
    int choice;
    string username, password;
    while (true) {
        cout << "Welcome to the User Authentication System" << endl;
        cout << "1. Register" << endl;
        cout << "2. Login" << endl;
        cout << "3. Exit" << endl;
        cout << "Enter your choice: ";
        cin >> choice;

        switch (choice) {
            case 1:
                cout << "Enter username: ";
                cin >> username;
                cout << "Enter password: ";
                cin >> password;
                registerUser(hStmt, username, password);
                break;
            case 2:
                cout << "Enter username: ";
                cin >> username;
                cout << "Enter password: ";
                cin >> password;
                if (loginUser(hStmt, username, password)) {
                    cout << "Login successful!" << endl;
                } else {
                    cout << "Login failed. Invalid username or password." << endl;
                }
                break;
            case 3:
                // 退出程序
                SQLDisconnect(hConn);
                SQLFreeHandle(SQL_HANDLE_DBC, hConn);
                SQLFreeHandle(SQL_HANDLE_ENV, hEnv);
                return 0;
            default:
                cout << "Invalid choice. Please enter a valid option." << endl;
        }
    }

    return 0;
}

// 用户注册函数
void registerUser(SQLHANDLE hStmt, const string& username, const string& password) {
    SQLCHAR query[255];
    SQLINTEGER cbUsername = SQL_NTS;
    SQLINTEGER cbPassword = SQL_NTS;

    printf_s((char*)query, "INSERT INTO Users (username, password) VALUES ('%s', '%s')", username.c_str(), password.c_str());

    SQLAllocHandle(SQL_HANDLE_STMT, hStmt, &hStmt);
    SQLExecDirect(hStmt, query, SQL_NTS);
    SQLFreeHandle(SQL_HANDLE_STMT, hStmt);

    cout << "User registered successfully!" << endl;
}

// 用户登录函数
bool loginUser(SQLHANDLE hStmt, const string& username, const string& password) {
    SQLCHAR query[255];
    SQLINTEGER cbUsername = SQL_NTS;
    SQLINTEGER cbPassword = SQL_NTS;

    printf_s((char*)query, "SELECT * FROM Users WHERE username='%s' AND password='%s'", username.c_str(), password.c_str());

    SQLAllocHandle(SQL_HANDLE_STMT, hStmt, &hStmt);
    SQLExecDirect(hStmt, query, SQL_NTS);

    SQLCHAR result[255];
    while (SQLFetch(hStmt) == SQL_SUCCESS) {
        SQLGetData(hStmt, 1, SQL_CHAR, result, sizeof(result), NULL);
        if (strcmp((char*)result, username.c_str()) == 0) {
            SQLFreeHandle(SQL_HANDLE_STMT, hStmt);
            return true;
        }
    }

    SQLFreeHandle(SQL_HANDLE_STMT, hStmt);
    return false;
}

File: 课程作业/Python/database/dbc.py
import pyodbc
import pymssql

# SQL Server连接信息
SQL_SERVER = 'LAPTOP-CYC\MSSQLSERVER02'
SQL_DB = 'TestDB'
SQL_USER = 'TestUser'
SQL_PASSWORD = 'test123456'

# 连接数据库
conn_str = f'DRIVER={{SQL Server}};SERVER={SQL_SERVER};DATABASE={SQL_DB};UID={SQL_USER};PWD={SQL_PASSWORD}'
conn = pyodbc.connect(conn_str)
# conn = pymssql.connect(Server=SQL_SERVER, Database=SQL_DB, Security=True)
cursor = conn.cursor()


# 用户注册函数
def register_user(username, password):
    try:
        cursor.execute("INSERT INTO Users (username, password) VALUES (?, ?)", (username, password))
        conn.commit()
        print("注册成功!")
    except pyodbc.Error as e:
        print("注册失败:", e)


# 不安全的用户登录（容易受到SQL注入攻击）
def unsafe_login(username, password):
    try:
        query = "SELECT * FROM Users WHERE 账号 = '{}' AND 密码 = '{}'".format(username, password)
        cursor.execute(query)
        row = cursor.fetchone()
        if row:
            print("不安全登录成功!")
        else:
            print("不安全登录失败!")
    except pyodbc.Error as e:
        print("不安全登录失败:", e)


# 安全的用户登录（使用参数化查询）
def safe_login(username, password):
    try:
        query = "SELECT * FROM Users WHERE 账号 = ? AND 密码 = ?"
        cursor.execute(query, (username, password))
        row = cursor.fetchone()
        if row:
            print("安全登录成功!")
        else:
            print("安全登录失败!")
    except pyodbc.Error as e:
        print("安全登录失败:", e)


# 用户界面
while True:
    print("欢迎使用本系统！请选择下列功能：")
    print("1. 注册")
    print("2. 不安全登录")
    print("3. 安全登录")
    print("4. 退出")

    choice = input("请选择操作：")

    if choice == '1':
        username = input("请输入账号：")
        password = input("请输入密码：")
        register_user(username, password)
    elif choice == '2':
        username = input("请输入账号：")
        password = input("请输入密码：")
        unsafe_login(username, password)
    elif choice == '3':
        username = input("请输入账号：")
        password = input("请输入密码：")
        safe_login(username, password)
    elif choice == '4':
        conn.close()
        print("再见!")
        break
    else:
        print("输入无效，请重新输入。")
